finetune/input_pipeline.py

```diff
@@ -0,0 +1,209 @@
+        for out in out_gen:
+            feats = {"tokens": out.token_ids, "mask": out.mask}
+            if Y is None:
+                yield feats
+            else:
+                yield feats, self.label_encoder.transform([Y])[0]
+
+    def _post_data_initialization(self, Y):
+        self.label_encoder = self._target_encoder()
+        if not callable(Y):
+            Y_fit = Y
+            self.config.dataset_size = len(Y)
+            self.label_encoder.fit_transform(Y)
+        else:
+            try:
+                self.config.dataset_size = len(Y())
+            except TypeError:
+                logging.warning(
+                    "Generator input function does not have a length, falling back to default in config of {}".format(
+                        self.config.dataset_size))
+            Y_fit = list(itertools.islice(Y(), 100))  # TODO find a more principled way to do this?
+            self.label_encoder.fit_transform(Y_fit)
+
+        target_dim = self.label_encoder.target_dim
+        self.lm_loss_coef = self.config.lm_loss_coef if target_dim is not None else 1.0
+        if target_dim != self.target_dim and self.target_dim is None:
+            self.rebuild = True
+        self.target_dim = target_dim
+
+        if Y_fit is not None:
+            self.config.class_weights = compute_class_weights(class_weights=self.config.class_weights, Y=Y_fit)
+
+    def _dataset_with_targets(self, Xs, Y):
+        if not callable(Xs) and not callable(Y):
+            dataset = lambda: zip(Xs, Y)
+        elif callable(Xs) and callable(Y):
+            dataset = lambda: zip(Xs(), Y())  # encode one sample at a time.
+        else:
+            raise ValueError("Either neither or both of Xs and Y should be callable, not a mixture")
+
+        dataset_encoded = lambda: itertools.chain.from_iterable(
+            map(lambda xy: self.text_to_tokens_mask(*xy), dataset()))
+        shape_def = self.feed_shape_type_def()
+        if len(self.config.visible_gpus) > 1:
+            dataset_encoded_ = dataset_encoded
+            dataset_encoded = lambda: map(self._adapt_feats_for_distributed, dataset_encoded_())
+            shape_def = [self._adapt_feats_for_distributed(s) for s in shape_def]
+        return Dataset.from_generator(dataset_encoded, *shape_def)
+
+    def _dataset_without_targets(self, Xs):
+        if not callable(Xs):
+            Xs_fn = lambda: Xs
+        else:
+            Xs_fn = Xs
+
+        dataset_encoded = lambda: itertools.chain.from_iterable(map(self.text_to_tokens_mask, Xs_fn()))
+        types, shapes = self.feed_shape_type_def()
+        return Dataset.from_generator(dataset_encoded, types[0], shapes[0])  # 0s cut out the targets
+
+    def _get_train_input_fns(self, Xs, Y=None, batch_size=None, val_size=None):
+        batch_size = batch_size or self.config.batch_size
+
+        shuffle_buffer_size = 100
+        val_size = val_size or 0
+        prefetch_buffer = 2  # breaks the pipeline to allow concurrency
```

Comment:
This feels like something that should live in configuration rather than as a hardcoded constant.